cleanrl/atari/v_c51_atari.py

```diff
@@ -0,0 +1,346 @@
+        self.network = nn.Sequential(
+            nn.Conv2d(frames, 32, 8, stride=4),
+            nn.ReLU(),
+            nn.Conv2d(32, 64, 4, stride=2),
+            nn.ReLU(),
+            nn.Conv2d(64, 64, 3, stride=1),
+            nn.ReLU(),
+            nn.Flatten(),
+            nn.Linear(3136, 512),
+            nn.ReLU(),
+            nn.Linear(512, envs.single_action_space.n * n_atoms)
+        )
+
+    def forward(self, x, device):
+        #x = torch.Tensor(x).to(device)
+        return self.network(x/255.0)
+
+    def get_action(self, envs, device, x,action=None):
+        logits = self.forward(x, device)
+        # probability mass function for each action
+        pmfs = torch.softmax(logits.view(len(x), envs.single_action_space.n, self.n_atoms), dim=2)
+        q_values = (pmfs*self.atoms).sum(2)
+        if action is None:
+            action = torch.argmax(q_values, 1)
+            random_action = torch.randint(0, envs.single_action_space.n, (len(x),), device=device)
+            random_action_flag = torch.rand(len(x), device=device) > epsilon
+            action = torch.where(random_action_flag, action, random_action)
+           
+        return action, pmfs[torch.arange(len(x)), action]
+
+def linear_schedule(start_e: float, end_e: float, duration: int, t: int):
+    slope =  (end_e - start_e) / duration
+    return max(slope * t + start_e, end_e)
+
+
+q_network = QNetwork(envs, device, n_atoms=args.n_atoms, v_min=args.v_min, v_max=args.v_max).to(device)
+target_network = QNetwork(envs, device, n_atoms=args.n_atoms, v_min=args.v_min, v_max=args.v_max).to(device)
+target_network.load_state_dict(q_network.state_dict())
+optimizer = optim.Adam(q_network.parameters(), lr=args.learning_rate, eps=0.01/args.batch_size)
+loss_fn = nn.MSELoss()
+print(device.__repr__())
+print(q_network)
+
+# ALGO Logic: Storage setup
+obs = torch.zeros((args.num_steps, args.num_envs) + envs.single_observation_space.shape).to(device)
+actions = torch.zeros((args.num_steps, args.num_envs) + envs.single_action_space.shape).to(device)
+rewards = torch.zeros((args.num_steps, args.num_envs)).to(device)
+dones = torch.zeros((args.num_steps, args.num_envs)).to(device)
+
+# TRY NOT TO MODIFY: start the game
+
+global_step = 0
+start_time = time.time()
+next_obs = torch.Tensor(envs.reset()).to(device)
+next_done = torch.zeros(args.num_envs).to(device)
+num_updates = args.total_timesteps // args.batch_size
+num_gradient_updates = 0
+
+for update in range(1, num_updates + 1):
+    # ROLLOUTS
+    epsilon = linear_schedule(args.start_e, args.end_e, args.exploration_fraction*args.total_timesteps, global_step)
+    writer.add_scalar("charts/epsilon", epsilon, global_step)
+    for step in range(0, args.num_steps):
+        global_step += 1 * args.num_envs
+        obs[step] = next_obs
+        dones[step] = next_done
+        with torch.no_grad():
+            action, _ = q_network.get_action(envs, device, next_obs)
```

Comment:
Where is the epsilon greedy exploration? In VDQN, I had [these lines](https://github.com/vwxyzjn/cleanrl/pull/59/files#diff-bfb142a3850419f93ee126b7e8ec527b4bca61221809062a7030e8f147685bd5R199-R206)

```python
            # ALGO LOGIC: action logic
            with torch.no_grad():
                logits = q_network.forward(next_obs)
                action = torch.argmax(logits, dim=1)
                random_action = torch.randint(0, envs.single_action_space.n, (envs.num_envs,), device=device)
                random_action_flag = torch.rand(envs.num_envs, device=device) > epsilon
                action = torch.where(random_action_flag, action, random_action)
            actions[step] = action
```

Reply:
The epsilon greedy exploration is in the Agents get_action method.

```
def get_action(self, envs, device, x,action=None):
        logits = self.forward(x, device)
        # probability mass function for each action
        pmfs = torch.softmax(logits.view(len(x), envs.single_action_space.n, self.n_atoms), dim=2)
        q_values = (pmfs*self.atoms).sum(2)
        if action is None:
            action = torch.argmax(q_values, 1)
            random_action = torch.randint(0, envs.single_action_space.n, (len(x),), device=device)
            random_action_flag = torch.rand(len(x), device=device) > epsilon
            action = torch.where(random_action_flag, action, random_action)

        return action, pmfs[torch.arange(len(x)), action]
```